Enable shared memory based RMA for win_allocate

minsii opened the pull request:
This PR enables shared memory for `win_allocate`. It includes following changes:
- Add shm symmetric heap allocation (use MPL_SHM*, collective, mmap)
- Use common logic for both `win_allocate` and `win_allocate_shared`
- Allocation logic: 
    * If global shm symmetric heap can be allocated (see limitation below), then try to allocate symheap 
        + The leading process generates a random start address
        + If multiple processes exist no a node, use MPL_SHM with a fixed start address
        + If only single process exists on a node, use `MPL_mmap` instead. 
        + If mapping failed, retry predefined times.
    * If symheap allocation failed or it is not supported, then allocate non-symmetric region
        + If multiple processes exist no a node, allocate shared memory region per node (MPL SHM + collective)
        + If only single process exists on a node, use `malloc` instead.

**SHM symheap limitation**: 
If shared memory segment is created within symheap (i.e., multiple pes no a node), the window size of each process must be page aligned, or (alloc_shared_noncontig=true) is set thus we can internally adjust the mapping size on each process. Otherwise, we can only allocate shared memory regions separately on each node (non-symmetric).

Here are an example of symheap and simple design description:
[ECP-win_allocate_shm_symm_heap.pptx](https://github.com/pmodels/mpich/files/2390295/ECP-win_allocate_shm_symm_heap.pptx)

**NOTES**
Please ignore the first two `mpl` commits, they are moved into #3335 but kept here only for jenkins tests.

minsii commented:
test:jenkins/ch4/most

minsii commented:
test:jenkins/ch4/most

minsii commented:
test:jenkins/ch4/most

minsii commented:
test:jenkins/ch4/most

minsii commented:
test:jenkins/ch4/most

minsii commented:
test:jenkins/ch4/most

minsii commented:
test:jenkins/ch4/most

minsii commented:
test:jenkins/ch4/most

minsii commented:
@pavanbalaji Are you available to review this PR ? The symheap design has be changed based on our discussion two weeks ago.

minsii commented:
The `ucx` failure seems irrelevant.

minsii commented:
Rebased agains latest master to exclude MPL commits from this PR.

hajimefu commented:
test:intel:ch4/ofi/sockets/gnu/force-am-disabled,force-am-enabled/auto,netmod/debug,default

hajimefu commented on src/mpid/ch4/src/ch4r_win.h:
```diff
@@ -1269,6 +1269,15 @@ static inline int MPIDI_CH4R_mpi_win_shared_query(MPIR_Win * win,
+        *((void **) baseptr) = win->base;
+        goto fn_exit;
+    }
+
```
This patch (`ch4/rma: fix win_shared_query when no shm env`) should be squashed with the appropriate patch before merging.

minsii replied:
Make sense. I will squash it with patch `ch4/rma: define win alloc common routine using shm symheap`, where `win_allocate` starts creating shared memory.

hajimefu commented on src/mpid/ch4/src/ch4r_symheap.h:
```diff
@@ -312,12 +312,11 @@ static inline int MPIDI_CH4R_get_shm_symheap(MPI_Aint shm_size, MPI_Aint * shm_o
 {
     int mpi_errno = MPI_SUCCESS;
     int iter = MPIR_CVAR_CH4_SHM_SYMHEAP_RETRY;
+    unsigned any_mapfail_flag = 1;
```
This fix ("ch4/rma: fix symheap failure checking") should be squashed with an appropriate patch before merging.

minsii replied:
will squashed

hajimefu commented on src/mpid/ch4/src/ch4r_win.h:
```diff
@@ -1117,22 +1118,36 @@ static inline int MPIDI_CH4I_win_shm_alloc_impl(MPI_Aint size,
     } else
         total_size = size;
 
+    /* global symm heap can be used only when any of the following conditions meets:
```
s/meets/meet/

hajimefu commented on src/mpid/ch4/src/ch4r_win.h:
```diff
@@ -1117,22 +1118,36 @@ static inline int MPIDI_CH4I_win_shm_alloc_impl(MPI_Aint size,
+    /* global symm heap can be used only when any of the following conditions meets:
+     * - no shared memory node (i.e., single process per node)
+     * - size of each process on the shared memory node is page aligned
+     * - user sets alloc_shared_noncontig=true (TODO) */
```
What was `alloc_shared_noncontig=true`? A new info key being proposed?

hajimefu replied:
Oh okay, I totally missed it. Sorry!

minsii commented on src/mpid/ch4/shm/posix/posix_win.h:
```diff
@@ -400,20 +400,44 @@ MPL_STATIC_INLINE_PREFIX int MPIDI_POSIX_mpi_win_create_hook(MPIR_Win * win)
+#undef FUNCNAME
+#define FUNCNAME MPIDI_POSIX_mpi_win_allocate_hook
+#undef FCNAME
+#define FCNAME MPL_QUOTE(FUNCNAME)
```
Emm, but the new code in  `MPIDI_POSIX_mpi_win_allocate_hook` will fail at compiling time if I do not include the `FUNCNAME ` & `FCNAME ` declaration.

hajimefu replied:
Then this FUNCNAME & FCNAME declaration can go first and the rest of the changes come after?

minsii replied:
Is it too redundant ? I guess having them in a single commit might make more sense.

hajimefu replied:
I'm just saying based on "one commit does one thing" policy. If this commit were to add missing `FUNCNAME`/`FCNAME`, it would have been fine.

minsii replied:
@hajimefu : Could you please let me know if you have any feedback on this problem ?

hajimefu replied:
Okay, I still don't 100% agree but I'll approve to have it in one patch.

minsii commented on src/mpid/ch4/src/ch4r_win.h:
```diff
@@ -855,25 +855,26 @@ static inline int MPIDI_CH4R_win_finalize(MPIR_Win ** win_ptr)
+            /* if single process on the node, we use mmap with symm heap */
             MPL_munmap(MPIDI_CH4U_WIN(win, mmap_addr), MPIDI_CH4U_WIN(win, mmap_sz), MPL_MEM_RMA);
-        else if (MPIDI_CH4U_WIN(win, mmap_sz) == -1)
+        } else if (win->base) {
```
Will fix

minsii replied:
fixed

minsii commented on src/mpid/ch4/src/ch4r_win.h:
```diff
@@ -1046,83 +1047,161 @@ static inline int MPIDI_CH4R_mpi_win_attach(MPIR_Win * win, void *base, MPI_Aint
 
-    /* allocate symmetric shared window memory */
-    size_t page_sz, mapsize;
+        if (total_size == 0 && shm_comm_ptr->local_size == comm_ptr->local_size)
```
It checks `if all processes give zero size`, and `if this window contains only a single node`. 
- If both true, then it just returns. 
- If the window contains several nodes, other processes on the remote node might or might not be zero size: 
  * If global symheap enabled: we need to handle it in `MPIDI_CH4R_get_shm_symheap`
  * if globa symheap disabled: local process just does nothing and return.

I will add a comment on top of the line.

minsii commented on src/mpid/ch4/src/ch4r_symheap.h:
```diff
@@ -193,6 +193,238 @@ static inline void *MPIDI_CH4R_generate_random_addr(size_t size)
+    if (mpi_errno != MPI_SUCCESS)
+        goto fn_fail;
+
+    if (maxloc_result.sz > 0) {
```
Sounds good. Will change it.

minsii commented:
test:jenkins/ch4/most

minsii commented:
@hajimefu : This push should fix most of your comments, except the one about `FUNCNAME` which we are still discussing..

minsii commented on src/mpid/ch4/src/ch4r_symheap.h:
```diff
@@ -193,6 +193,236 @@ static inline void *MPIDI_CH4R_generate_random_addr(size_t size)
+    MPIR_Comm *shm_comm_ptr = comm->node_comm;
+
+    struct {
+        size_t sz;
```
I just reused the code from old symheap allocation. But you are right that the types seems not match. How about replace it with `MPI_Aint` ?

hajimefu replied:
Also, if you are to change this to `long`, you'll need to be careful about the code like this.
https://github.com/pmodels/mpich/pull/3319/files/101f448325a0e15acff2e8c3efc7831d3a3d9bc1..d7faf7a4119d2f18593f152a0cd437fda14638c9#diff-4a7415a94d2f7728e88418ad6a750a76R336

```C
maxloc.sz = mapsize;
```
will no longer be correct because you may be truncating the value.

Another idea would be to define a user-defined operator that supports `MAXLOC` for `size_t`?

minsii replied:
We might want to change the type of all dependent parameters. `MPI_Aint` represents address value, which should be safer than `long`. It can be also directly translated to `MPI_AINT` in the reduce operation.

minsii commented on src/mpid/ch4/src/ch4r_win.h:
```diff
@@ -1240,13 +1328,13 @@ static inline int MPIDI_CH4R_mpi_win_allocate(MPI_Aint size,
 
     win = *win_ptr;
     win->base = baseP;
+    win->size = size;
```
It was set in `MPIDI_CH4R_get_symmetric_heap`, which was inappropriate in my opinion. Because the symheap function should not set window specific attribute. Although I still pass `win` input in the new symheap function, it only updates symheap related parameters. 

In this commit, old `MPIDI_CH4R_get_symmetric_heap` is replaced with new `MPIDI_CH4I_win_shm_alloc_impl`, thus I need to add the size setting here.

minsii commented:
Maybe it is better to replace `total_size` with `total_shm_size` so it is clear. 

About global symheap conditions:
* size of each process on the shared memory node is page aligned
Because most memory mapping routine (e.g., `mmap`) requires page aligned address, the picture in the attached ppt file might be helpful (see top of this PR)

* user sets alloc_shared_noncontig=true
Because no matter what size is specified, we can internally change it if user allows noncontig shared memory. So it can be page aligned.

I think it will be useful if I make the code comment more detailed.

About the commit logic, I actually wanted to make sure the global symheap logic works correctly even the above condition does not meet. That is, it retries 100 times and releases resource correctly, and then fallback to a backup approach. Adding conditions to disable it can be considered as an optimization to avoid unnecessary retry. I may also adjust the comment to make the logic clear. Does it make sense to you ?

hajimefu commented:
> Maybe it is better to replace total_size with total_shm_size so it is clear.

Sounds good to me.

> About global symheap conditions

I see. Thank you for explanations.

> About the commit logic

Ah okay. I was not aware of that logic. Now it makes sense.

minsii commented:
This push improves comments and `total_size `->`total_shm_size`. But it does not address the maxloc  with size_t issue. I am going to implement user defined op.

minsii commented:
test:jenkins/ch4/most

minsii commented:
test:jenkins/ch4/most

minsii commented:
@hajimefu : I added user-defined op and datatype to handle size_t with maxloc reduce. The changes are included in two commits: 
- [#52198d5](https://github.com/pmodels/mpich/pull/3319/commits/52198d58940af7fce58fd1bba23c82e4f0b76a14) mpi/coll: define MPIR_Op_create|free_impl
- [#245d2ae](https://github.com/pmodels/mpich/pull/3319/commits/245d2ae63466880bc6d9ba7390b97abdc5a44189) ch4/rma: implement symmetric heap with shm seg

Besides, I also made some minor changes:
- In [#a6c2264](https://github.com/pmodels/mpich/pull/3319/commits/a6c22648c88a86f7130a75cbf068cd5f0486bdf0) ch4/rma: remove unused old symm heap
   Also deleted the old CVAR
- In [#5b87f6f](https://github.com/pmodels/mpich/pull/3319/commits/5b87f6fd4b2962612b51be997af769aaa76c3918) ch4/rma: enable global symheap only when needed
  Skip global sym heap when only 1 process exists in the window

Could you please review the changes for me ?

hajimefu commented on src/mpid/ch4/src/ch4r_symheap.h:
```diff
@@ -193,6 +193,309 @@ static inline void *MPIDI_CH4R_generate_random_addr(size_t size)
+    }
+}
+
+/* Allredue MAXLOX for unsigned size type by using user defined operator
```
Allreduce MAXLOC

Also let's mention why we needed this i.e. standard `MAXLOC` doesn't support unsigned types, thus not compatible with `size_t`.

minsii replied:
Sounds good to me. I just pushed the improved comment. Hopefully it does not interrupt the tests.

hajimefu commented:
Changes look good. User-defined MAXLOC is complicated but I think this is the best we can. I'll hit the approve button once I confirm the test result from our system.

hajimefu commented:
test:intel:ch4/ofi/sockets/gnu/force-am-disabled,force-am-enabled/auto,netmod,no-odd-even/debug,default
